---
title: "Selection test"
author: "Lorena Lorenzo FernÃ¡ndez"
date: "25 March, 2021"
output:
csl: 
editor_options:
  chunk_output_type: console
---

## 1. Rehh, an R package for analysing iHS and XP-EHH

### 1.1 Prepare data

Once I have filtered, phased and polarized data (see variant_filtering.md , phasing.md or polarization.md), I can start with selection analyses. The vcf file containing such information is:

`${sp}_goodsamples_filtered_phased_polarized_variants_header_cat_ref.vcf`

After finding the error in the number variants, I had to run again selection analyses. As CSIC is under informatic attack, I don't have VPN access to genomic servers, so I am repeating the analysis in CESGA server. That's why I am changing parts of the code.

```{bash}
#Separate per chr sp vcf
species=(lc ll lp lr)

for sp in ${species[@]}
do
  echo "$sp"
  CHR=($(cat $STORE2/reference_genomes/Felis_catus_Ref/Felis_catus.Felis_catus_9.0.dna.toplevel.fa.fai | cut -f 1 | uniq))
  for chr in ${CHR[@]:0:19}
    do
      echo "$chr"
      grep -E "^(#|${chr})" \
      $LUSTRE/${sp}_goodsamples_filtered_phased_polarized_variants_header_cat_ref.vcf \
      > $LUSTRE/${chr}_${sp}_goodsamples_filtered_phased_polarized_variants_header_cat_ref.vcf
      echo "Done $chr for $sp vcf"
    done
done
```

### 1.2 iHS analyses

For iHS analyses we decided to use the "rehh" package available for R. The entire code is available in ihs_sp.R. To run this in CESGA, I had to create a run_ihs.sh script that specify to run the Rscript.

NOTE: We found that Kovalaskas et al. 2020 used an adjusted p-value to correct for False Discovery Rate (FDR) using Benjamini-Hochberg method. After investigating about the method and the meaning I tried to integrate in my analyses. The option has to be marked in the ihh2ihs function (See ihs_sp.R)

### 1.3 XP-EHH analyses

Discuss here (relying in previous studies) the difference btw this methods, what should we use and why. Taken from Talenti2017:

*The Rsb statistic compares EHH for the same SNP in two different populations and can provide evidence of selection given the presence of high-frequency or fixed alleles in one population but not on the other (Tang et al. 2007). Similarly, the XP-EHH detected selective sweeps in which one allele had undergone strong directional selection in one population while remaining polymorphic in the population as a whole (Sabeti et al. 2007)*

We finally decided to use XP-EHH (See xpehh_sp.R). As with iHS, I used rehh R package. To run this in CESGA, I had to create a run_xpehh.sh script that specify to run the Rscript.

## 2. Lassip, a program for analysing HFS

```{bash}

species=(lc ll lp lr)

for sp in ${species[@]}
  do
    echo "$sp"
    #make a sp_ind.txt (pop file) needed for lassip
    sed "s/$/\t${sp}/" <(grep "#CHR" $LUSTRE/${sp}_goodsamples_filtered_phased_polarized_variants_header_cat_ref.vcf | tr "\t" "\n" | grep ${sp})  > $LUSTRE/selection_scan/${sp}_ind.txt
    sbatch lassi.sh $sp
  done  
```

## 3. Ploting results

See plot_ihs_lassi.R

After this, I want to merge every overlapping outlier window with bedtools merge:

```{bash}
species=(lc ll lp lr)

for sp in ${species[@]}
  do
    echo "$sp"
    bedtools merge -i ${sp}_99outliers.bed -c 1 -o count \
    > ${sp}_99outliers_merged.bed
  done  
```

### 3.1 Intersect between the two methods

```{bash}
species=(lc ll lp lr)

for sp in ${species[@]}
  do
    echo "$sp"
    bedtools intersect \
    -a /home/llorenzo/selection_scan/lassi_test/${sp}_99outliers_merged.bed \
    -b /home/llorenzo/selection_scan/rehh_test/${sp}_ihs_99.9outliers.bed \
    -wao \
    > /home/llorenzo/selection_scan/${sp}_outliers_ihslassi_intersect.bed
  done  
```

Once did this I am going to plot shaded areas from LASSI outliers over the iHS plot results

```{r}

#Name the variables
species<- c("lc", "ll", "lp", "lr")


for (sp in species)
{
  ###LASSI###
  #Read the data
  path <- "/home/llorenzo/selection_scan/lassi_test/"
  lassi_outliers <- read.table(paste0(path, sp, "_99outliers_merged.bed"), sep="\t")
  colnames(lassi_outliers) <- c("chr", "start", "end", "n_windows")
    
  #Group by chr
  data_lassi_cum <- lassi_outliers %>% 
    group_by(chr) %>% 
    summarise(max_bp = max(end)) %>%
    mutate(bp_add = lag(cumsum(max_bp), default = 0)) %>% 
    select(chr, bp_add)

  data_lassi <- lassi_outliers %>% 
    inner_join(data_lassi_cum, by = "chr") %>% 
    mutate(bp_start_cum = start + bp_add) %>%
    mutate(bp_end_cum = end + bp_add) %>%
    na.omit()
  
  ###iHS###
  #Read the data
  path <- "/home/llorenzo/selection_scan/rehh_test/"
  ihs_scan <- read.table(paste0(path, sp, "_ihs_scan"), sep="\t", header= TRUE)

  #Group by chr
  data_ihs_cum <- ihs_scan %>% 
    group_by(CHR) %>% 
    summarise(max_bp = max(POSITION)) %>%
    mutate(bp_add = lag(cumsum(max_bp), default = 0)) %>% 
    select(CHR, bp_add)

  data_ihs <- ihs_scan %>% 
    inner_join(data_ihs_cum, by = "CHR") %>% 
    mutate(bp_cum = POSITION + bp_add) %>%
    na.omit()

  ihs_axis_set <- data_ihs %>% 
    group_by(CHR) %>% 
    summarize(center = mean(bp_cum))

  #get the 0.1% of data
  cut_99.9 <- quantile(data_ihs$LOGPVALUE,probs=0.999,na.rm=T)
  
  print(paste(sp, round(length(data_ihs$LOGPVALUE) * 0.001), "top 0.1% values"))
  print(paste(sp, cut_99.9, "0.1% cut-off" ))
  
  print("Cut-offs done")
  
#Plot ihs with lassi windows  
path<- "/home/llorenzo/selection_scan/"
pdf(paste0(path, sp, "_ihs_lassi_cutoffs_manhattanplot.pdf"), width=10, height=5)

print(ggplot() +
        geom_point(data_ihs, mapping=aes(x = bp_cum, y = LOGPVALUE, color = as_factor(CHR)), alpha = 0.5) +
        geom_hline(yintercept= cut_99.9, color="grey", linetype="dashed") +
        geom_rect(data_lassi, mapping=aes(xmin = bp_start_cum, xmax = bp_end_cum, ymin = -Inf, ymax = Inf), alpha = 0.4, color = as_factor(CHR)) +
        scale_x_continuous(label = axis_set$CHR, breaks = axis_set$center) +
        labs (x="Chromosome", y="p-adjust iHS", title=sp ) +
        theme(plot.title = element_text(hjust = 1, vjust = -10), legend.position = "none"))

dev.off()

print("Plot done and saved")
}
```

Any conclusion can be extracted from this plot. I represented all SNPs from iHS and windows can rarely be seen as they are really small in contrast with the chr length, so as a result, we can only see a vertical line representing the window and lots of points representing SNPs.

So, the next thing I am going to try is to plot per chr outliers windows from saltilassi and the outliers SNPs from iHS.

```{r}
#Name the variables
species<- c("lc", "ll", "lp", "lr")


for (sp in species)
{
  ###LASSI###
  #Read the data
  path <- "/home/llorenzo/selection_scan/lassi_test/"
  lassi_outliers <- read.table(paste0(path, sp, "_99outliers_merged.bed"), sep="\t") 
  colnames(lassi_outliers) <- c("CHR", "START", "END", "N_WINDOWS")
  lassi_outliers <-  filter(lassi_outliers, CHR != "X")
  print("Reading lassi results done")
  
  ###iHS###
  #Read the data
  path <- "/home/llorenzo/selection_scan/rehh_test/"
  ihs_outliers <- read.table(paste0(path, sp, "_ihs_99.9outliers"), sep="\t", header= TRUE)
  colnames(ihs_outliers) <- c("CHR", "POS", "IHS", "PVAL")
  print("Reading iHS results done")
  
  #Plot ihs with lassi windows  
#Define chromosome variable
chromosomes <- c("A1", "A2", "A3", "B1", "B2", "B3", "B4", "C1", "C2", "D1", "D2", "D3", "D4", "E1", "E2", "E3", "F1", "F2")

#loop per chr plot
for (chr in chromosomes)
{
chr_data_ihs <- ihs_outliers %>%
  filter(CHR== chr)
chr_data_lassi <- lassi_outliers %>%
  filter(CHR== chr)
    
path<- "/home/llorenzo/selection_scan/"
pdf(paste0(path, chr, "_", sp, "_ihs_lassi_manhattanplot.pdf"), width=10, height=5)

print(ggplot() +
        geom_point(chr_data_ihs, mapping=aes(x = POS, y = PVAL, alpha = 0.5)) +
        geom_rect(chr_data_lassi, mapping=aes(xmin = START, xmax = END, ymin = -Inf, ymax = Inf), alpha = 0.4) +
        labs (x="Chromosome", y="p-adjust iHS", title=sp ) +
        theme(plot.title = element_text(hjust = 1, vjust = -10), legend.position = "none"))

dev.off()

print("Plot done and saved")
}
}
```

By doing this I discovered that 0.1% of SNPS and 1% of windows is very restrictive. I am going to try with 1% of SNPS and 5% of windows. Well maybe is only in lc

PROBLEM FOUND: Lp has aprx. 8.2M variants, which is much more than expected... I went back to the sp_filter8.vcf and discover that by this point lp has 1.2M variants which is far more feasible. Now I have to traceback where I did it wrong (but unluckily this week I don't have access to CESGA, due to CSIC-cibersecurity problems)

### 3.3.1 PCA before and after phasing

```{bash}
###PCA
#Repeat plink per sp.
species=(lc ll lp lr)

for i in ${species[@]}
  do
    VCF=/home/llorenzo/selection_scan/rehh_test/${i}_goodsamples_filtered_phased_polarized_ACfiltered_header_cat_ref.vcf
    # perform linkage pruning - i.e. identify prune sites
    plink_1.9 --vcf $VCF --double-id --allow-extra-chr \
    --set-missing-var-ids @:# \
    --indep-pairwise 50 10 0.1 --out ${i}_goodsamples_filtered_phased_polarized_cat_ref
    #run PCA in the subset
    plink_1.9 --vcf $VCF --double-id --allow-extra-chr --set-missing-var-ids @:# \
    --extract ${i}_goodsamples_filtered_phased_polarized_cat_ref.prune.in \
    --make-bed --pca --out ${i}_goodsamples_filtered_phased_polarized_cat_ref
  done


```

```{r}
library(tidyverse)
library(viridis)


species<- c("lc", "ll", "lp", "lr")
path <- "~/"

##PREPHASED PCAs##
for (sp in species)
{
  #Read pre-phased data
  # import eigen vec and val
pca <- read_table(paste0(path, sp, "_goodsamples_cat_ref.eigenvec"),
                   col_names = FALSE)
eigenval <- scan(paste0(path, sp, "_goodsamples_cat_ref.eigenval"))  

# remove nuisance column
pca <- pca[,-1]
# set names
names(pca)[1] <- "ind"
names(pca)[2:ncol(pca)] <- paste0("PC", 1:(ncol(pca)-1))

print("named cols")

# first convert to percentage variance explained
pve <- data.frame(PC = 1:(ncol(pca)-1), pve = eigenval/sum(eigenval)*100)
# then make a plot
pdf(paste0(path, sp, "prephased_pve.pdf"))
print(ggplot(pve, aes(PC, pve)) +
        geom_bar(stat = "identity") +
        ylab("Percentage variance explained") + 
        theme_light())
dev.off()
print("pve plot done")

# plot PC1 PC2
pdf(paste0(path, sp, "prephased_pca.pdf"))

print(ggplot() +
   geom_point(pca, mapping=aes(x= PC1, y= PC2, color=ind) , size=4, alpha = 0.5) +
    geom_text(pca, mapping=aes(x= PC1, y= PC2, label=ind), size=3, check_overlap = TRUE) +
    coord_equal() +
    theme() +
    xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) +
    ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)")))
dev.off()
print("pca plot done")
}

##PHASED PCAs##
for (sp in species)
{
#Read phased data
pca <- read_table(paste0(path, sp, "_goodsamples_filtered_phased_polarized_cat_ref.eigenvec"),
                   col_names = FALSE)
eigenval <- scan(paste0(path, sp, "_goodsamples_filtered_phased_polarized_cat_ref.eigenval"))

print("read data")
# remove nuisance column
pca <- pca[,-1]
# set names
names(pca)[1] <- "ind"
names(pca)[2:ncol(pca)] <- paste0("PC", 1:(ncol(pca)-1))

print("named cols")

# first convert to percentage variance explained
pve <- data.frame(PC = 1:(ncol(pca)-1), pve = eigenval/sum(eigenval)*100)
# then make a plot
pdf(paste0(path, sp, "phased_pve.pdf"))
print(ggplot(pve, aes(PC, pve)) +
        geom_bar(stat = "identity") +
        ylab("Percentage variance explained") + 
        theme_light())
dev.off()
print("pve plot done")

# plot PC1 PC2
pdf(paste0(path, sp, "phased_pca.pdf"))

print(ggplot() +
   geom_point(pca, mapping=aes(x= PC1, y= PC2, color=ind) , size=4, alpha = 0.5) +
    geom_text(pca, mapping=aes(x= PC1, y= PC2, label=ind), size=3, check_overlap = TRUE) +
    coord_equal() +
    theme() +
    xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) +
    ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)")))
dev.off()
print("pca plot done")
}
```

```{bash}
bcftools view -e 'INFO/AF=1.00 | INFO/AF=0.00' *_lp_whatshap_phased_duplicated.vcf | wc -l
```

```{bash}

#LOOKING FOR THE ERROR

species=(lc lp lr)
CHR=($(cat $STORE2/reference_genomes/Felis_catus_Ref/Felis_catus.Felis_catus_9.0.dna.toplevel.fa.fai | cut -f 1 | uniq))

for sp in ${species[@]}
  do
    for chr in ${CHR[@]:0:19} 
        do
          grep -v "#" ${chr}_${sp}_whatshap_phased_duplicated.vcf  >> ${sp}_tryal_nvariants.vcf
      done
      wc -l ${sp}_tryal_nvariants.vcf
      cat <(grep "#" A1_${sp}_whatshap_phased_duplicated.vcf) ${sp}_tryal_nvariants.vcf > ${sp}_tryal_nvariants_header.vcf    
      bcftools view -e 'INFO/AF=1.00 | INFO/AF=0.00' ${sp}_tryal_nvariants_header.vcf | wc -l
    done
    

for sp in ${species[@]}
  do
    for chr in ${CHR[@]:0:19} 
        do
          grep -v "#" ${chr}_${sp}_shapeit_phased_duplicated_cat_ref.vcf  >> ${sp}_tryal_shapeit_nvariants.vcf
      done
      wc -l ${sp}_tryal_shapeit_nvariants.vcf
      cat <(grep "#" $LUSTRE/phasing/intermediate_files/A1_${sp}_whatshap_phased_duplicated.vcf) ${sp}_tryal_shapeit_nvariants.vcf > ${sp}_tryal_shapeit_nvariants_header.vcf    
      bcftools view -e 'INFO/AF=1.00 | INFO/AF=0.00' ${sp}_tryal_shapeit_nvariants_header.vcf | wc -l
    done  
   
```
